Alex and Happy Part 1/article.md

```diff
@@ -0,0 +1,817 @@
+
+Making a lexer using parser combinators is pretty doable and manageable.
+Meanwhile, using Alex is a bit more involved, but it has the advantage that performance will be more predictable; besides that, it can be easily integrated with Happy to output one token at a time, which may avoid creating a massive list of tokens in the memory in the first place.
+
+## Creating the project
+
+We will use Stack to build the project.
+Alex and Happy use files with extensions `.x` and `.y,` respectively, and Stack can automatically detect these file formats and build them.
+
+n.b.: Using Stack is not strictly necessary. You may also use Cabal or plain Alex and Happy files, and it will work as well.
+
+Begin by creating an empty Stack project:
+
+```bash
+$ stack new mini-ml simple
+$ cd mini-ml
+```
+
+You will also need to have Alex and Happy installed on your machine:
+
+```bash
+$ stack install alex
+$ stack install happy
+```
+
+We will create a `package.yaml` file for our project configuration.
+Use your favorite text editor to initialize `package.yaml` with the following:
+
+```yaml
+name: mini-ml
+version: 0.1.0.0
+
+extra-source-files:
+- README.md
+- LICENSE
+
+default-extensions:
+- OverloadedStrings
+
+library:
+  source-dirs: src
+  dependencies:
+  - array
+  - base
+  - bytestring
+  build-tools:
+  - alex
+  - happy
+```
+
+n.b.: If you chose to use Stack, you might need to use `hpack --force` to override the current Cabal file.
+
+You may want to remove `src/Main.hs`, as we won't need it.
+
+## Alex
+
+<!--
+* [X] link to its user guide
+-->
+
+Alex is a Haskell tool to generate lexers.
+
+This section aims at giving an introduction to Alex and how you can use it to do useful things in conjunction with Happy (explained later).
+Although Alex and Happy are frequently used together, they are independent tools. They may be combined with other technologies, so you may, for example, use Alex and Megaparsec together if you prefer.
+
+### How it works
+
+<!--
+* [X] make a nifty flowchart
+-->
+
+It will be essential to differentiate between the terms **lexeme** and **token** during this section.
+A lexeme is a valid atom of the grammar, such as a keyword (`in`, `if`, etc.), an operator (`+`, `/`, etc.), an integer literal, a string literal, a left or right parenthesis, an identifier, etc.
+Meanwhile, a token consists of the _token name_ and an optional _token value_.
+The token name is the name of a lexical category that the lexeme belongs to, while the token value is implemention-defined.
+We can represent tokens as a Haskell sum datatype, where data constructor names correspond to token names and constructor arguments correspond to the token value with optionally the scanned lexeme. For instance, `In`, `If`, `Plus`, `Divide`, `Integer 42`, `String "\"foo\""`, `LPar`, `RPar`, and `Identifier "my_function"` are all tokens generated for their corresponding lexemes.
+
```

Comment:
Seems like you would want some motivation sprinkled on top of here, something goal oriented. 

E.g.

"If your goal is to move from lexemes to tokens, how do you discern different lexemes? For that we use a DFA. "